src/survey/survey_info.rs

```diff
@@ -0,0 +1,187 @@
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+use std::fmt::Debug;
+
+use anyhow::Context;
+use netlink_packet_utils::{
+    nla::{DefaultNla, Nla, NlaBuffer},
+    parsers::{parse_u32, parse_u64},
+    DecodeError, Emitable, Parseable,
+};
+
+use crate::bytes::{write_u32, write_u64};
+
+const NL80211_SURVEY_INFO_FREQUENCY: u16 = 1;
+const NL80211_SURVEY_INFO_NOISE: u16 = 2;
+// const NL80211_SURVEY_INFO_IN_USE: u16 = 3;
+const NL80211_SURVEY_INFO_TIME: u16 = 4;
+const NL80211_SURVEY_INFO_TIME_BUSY: u16 = 5;
+const NL80211_SURVEY_INFO_TIME_EXT_BUSY: u16 = 6;
+const NL80211_SURVEY_INFO_TIME_RX: u16 = 7;
+const NL80211_SURVEY_INFO_TIME_TX: u16 = 8;
+// const NL80211_SURVEY_INFO_TIME_SCAN: u16 = 9;
+// const NL80211_SURVEY_INFO_PAD: u16 = 10;
+// const NL80211_SURVEY_INFO_TIME_BSS_RX: u16 = 11;
+// const NL80211_SURVEY_INFO_FREQUENCY_OFFSET: u16 = 12;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub enum Nl80211SurveyInfo {
+    // Center frequency of channel in MHz
+    Frequency(u32),
+    /// Noise level of channel (dBm)
+    Noise(i8),
+    /// Amount of time (in ms) that the radio was turned on (on channel or
+    /// globally)
+    ActiveTime(u64),
+    /// Amount of the time the primary channel was sensed busy (either due to
+    /// activity or energy detect)
+    BusyTime(u64),
+    /// Amount of time the extension channel was sensed busy
+    ExtensionBusyTime(u64),
+    /// Amount of time the radio spent receiving data (on channel or globally)
+    TimeRx(u64),
+    /// Amount of time the radio spent transmitting data (on channel or
+    /// globally)
+    TimeTx(u64),
+    Other(DefaultNla),
+}
+
+impl Nla for Nl80211SurveyInfo {
+    fn value_len(&self) -> usize {
+        match self {
+            Self::Frequency(v) => std::mem::size_of_val(v),
+            Self::Noise(v) => std::mem::size_of_val(v),
+            Self::ActiveTime(v)
+            | Self::BusyTime(v)
+            | Self::ExtensionBusyTime(v)
+            | Self::TimeRx(v)
+            | Self::TimeTx(v) => std::mem::size_of_val(v),
+            Self::Other(attr) => attr.value_len(),
+        }
+    }
+
+    fn kind(&self) -> u16 {
+        match self {
+            Self::Frequency(_) => NL80211_SURVEY_INFO_FREQUENCY,
+            Self::Noise(_) => NL80211_SURVEY_INFO_NOISE,
+            Self::ActiveTime(_) => NL80211_SURVEY_INFO_TIME,
+            Self::BusyTime(_) => NL80211_SURVEY_INFO_TIME_BUSY,
+            Self::ExtensionBusyTime(_) => NL80211_SURVEY_INFO_TIME_EXT_BUSY,
+            Self::TimeRx(_) => NL80211_SURVEY_INFO_TIME_RX,
+            Self::TimeTx(_) => NL80211_SURVEY_INFO_TIME_TX,
+            Self::Other(attr) => attr.kind(),
+        }
+    }
+
+    fn emit_value(&self, buffer: &mut [u8]) {
+        #[allow(clippy::cast_sign_loss)]
+        match self {
+            Self::Frequency(d) => write_u32(buffer, *d),
+            Self::Noise(d) => buffer[0] = *d as u8,
```

Comment:
`i8::cast_unsigned()`

Reply:
This seems to be a nightly feature https://doc.rust-lang.org/std/primitive.u8.html#method.cast_signed/, should I still use it? I'd prefer not to use nightly features, if no other nightly feature is used yet.

Reply:
Actually, this was [just stabilized](https://doc.rust-lang.org/stable/std/primitive.usize.html#method.cast_signed), so I'll use it

Reply:
I've used `cast_signed` and `uncast_signed` instead of the `as` cast in the latest revision.

Reply:
I changed it back because this bump minimum rust version to 1.87 which is not acceptable for old Linux distribution as RHEL/CentOS 9.